crates/codegen/parser/runtime/src/cursor.rs

```diff
@@ -6,41 +8,88 @@ use super::{
     text_index::{TextIndex, TextRange},
 };
 
+/// A [`NodePtr`] that points to a [`RuleNode`].
 #[derive(Clone, Debug, PartialEq, Eq)]
-struct CursorPathElement {
+struct RuleNodePtr {
     rule_node: Rc<RuleNode>,
     child_number: usize,
     text_offset: TextIndex,
 }
 
+impl RuleNodePtr {
+    fn into_node_ptr(self) -> NodePtr {
+        NodePtr {
+            node: Node::Rule(self.rule_node),
+            child_number: self.child_number,
+            text_offset: self.text_offset,
+        }
+    }
+}
+
+/// A pointer to a [`Node`] in a CST, used by the [`Cursor`] to implement the traversal.
 #[derive(Clone, Debug, PartialEq, Eq)]
-struct CursorLeaf {
```

Comment:
ditto

Reply:
It's the leaf of the path. So in alignment with the other name, this should be `LeafNode`

Reply:
Are we sure about that the naming here? In the context of traversing the tree, referring to the (current) final path element as leaf is very confusing to me, since leaf nodes (almost?) exclusively refer to the (terminal) nodes without children and the currently pointed to node is almost as probable to be an internal node as it is to be an outer one during the traversal.

Reply:
`PathLeafNode`? or `PathEndNode` or `Last` or `Final`. It's the principle I care about.

Reply:
Done in 03d055a13bd4ce7e4127442a290a766538597b2c